packages/contracts/contracts/src/AdvancedPolicy.sol

```diff
@@ -0,0 +1,103 @@
+    bool public immutable SKIP_POST;
+
+    /// @notice Controls whether main check can be executed multiple times.
+    bool public immutable ALLOW_MULTIPLE_MAIN;
+
+    /// @notice Tracks validation status for each subject per target.
+    /// @dev Maps target => subject => CheckStatus.
+    mapping(address => mapping(address => CheckStatus)) public enforced;
+
+    /// @notice Initializes contract with an AdvancedChecker instance and checks configs.
+    /// @param _advancedChecker Address of the AdvancedChecker contract.
+    /// @param _skipPre Skip pre-condition validation.
+    /// @param _skipPost Skip post-condition validation.
+    /// @param _allowMultipleMain Allow multiple main validations.
+    constructor(AdvancedChecker _advancedChecker, bool _skipPre, bool _skipPost, bool _allowMultipleMain) {
+        ADVANCED_CHECKER = _advancedChecker;
+        SKIP_PRE = _skipPre;
+        SKIP_POST = _skipPost;
+        ALLOW_MULTIPLE_MAIN = _allowMultipleMain;
+    }
+
+    /// @notice Enforces policy check for a subject.
+    /// @dev Only callable by target contract.
+    /// @param subject Address to validate.
+    /// @param evidence Validation data.
+    /// @param checkType Type of check (PRE, MAIN, POST).
+    function enforce(address subject, bytes calldata evidence, Check checkType) external override onlyTarget {
+        _enforce(subject, evidence, checkType);
+    }
+
+    /// @notice Internal check enforcement logic.
+    /// @dev Handles different check types and their dependencies.
+    /// @param subject Address to validate.
+    /// @param evidence Validation data.
+    /// @param checkType Type of check to perform.
+    /// @custom:throws CannotPreCheckWhenSkipped If PRE check attempted when skipped.
+    /// @custom:throws CannotPostCheckWhenSkipped If POST check attempted when skipped.
+    /// @custom:throws UnsuccessfulCheck If validation fails.
+    /// @custom:throws AlreadyEnforced If check was already completed.
+    /// @custom:throws PreCheckNotEnforced If PRE check is required but not done.
+    /// @custom:throws MainCheckNotEnforced If MAIN check is required but not done.
+    /// @custom:throws MainCheckAlreadyEnforced If multiple MAIN checks not allowed.
+    function _enforce(address subject, bytes calldata evidence, Check checkType) internal {
+        if (!ADVANCED_CHECKER.check(subject, evidence, checkType)) {
+            revert UnsuccessfulCheck();
+        }
+
+        CheckStatus storage status = enforced[msg.sender][subject];
+
+        // Handle PRE check.
+        if (checkType == Check.PRE) {
+            if (SKIP_PRE) revert CannotPreCheckWhenSkipped();
+            if (status.pre) {
+                revert AlreadyEnforced();
+            }
+
+            status.pre = true;
+        } else if (checkType == Check.POST) {
+            // Handle POST check.
+            if (SKIP_POST) revert CannotPostCheckWhenSkipped();
+            if (!status.pre) {
```

Comment:
maybe this is not needed? cause we check that we did main check, but in main check we never check that we did pre check, so it seems like the pre check becomes mandatory if you want a post check, but not if you just want a main check.

Reply:
or maybe you are missing if(!SKIP_PRE in the condition?

Reply:
right, I'm missing the `if(!SKIP_PRE && !status.pre)` in the condition. good catch!

Reply:
right, I'm missing the `if(!SKIP_PRE && !status.pre)` in the condition. good catch!

Reply:
actually there is no need for it, cause you have it in check main and before check post you need check main

Reply:
right, if actually the `pre` is skipped, it'll be catched in the `main` since is mandatory before doing `post` - I'll get rid of this line then